crates/ironrdp-connector/src/credssp_sequence.rs

```diff
@@ -0,0 +1,214 @@
+
+        let credssp_config: Box<dyn ProtocolConfig>;
+        if let Some(ref krb_config) = config.sspi_config {
+            credssp_config = Box::new(Into::<KerberosConfig>::into(krb_config.clone()));
+        } else {
+            credssp_config = Box::<sspi::ntlm::NtlmConfig>::default();
+        }
+        info!("using config : {:?}", &credssp_config);
+
+        let client = credssp::CredSspClient::new(
+            server_public_key,
+            credentials.into(),
+            credssp::CredSspMode::WithCredentials,
+            credssp::ClientMode::Negotiate(sspi::NegotiateConfig {
+                protocol_config: credssp_config,
+                package_list: None,
+                hostname: server_name,
+            }),
+            service_principal_name,
+        )
+        .map_err(|e| ConnectorError::new("CredSSP", ConnectorErrorKind::Credssp(e)))?;
+
+        match connector.state {
+            ClientConnectorState::CredSsp { selected_protocol } => Ok(Self {
+                client,
+                next_request: Some(credssp::TsRequest::default()),
+                state: CredSSPState::CredsspInitial,
+                selected_protocol,
+            }),
+            _ => Err(general_err!(
+                "Cannot perform cred ssp opeartions when ClientConnector is not in CredSsp state"
+            )),
+        }
+    }
+
+    pub fn is_done(&self) -> bool {
+        self.state == CredSSPState::Finishied
+    }
+
+    pub fn wants_request_from_server(&self) -> bool {
+        self.next_request.is_none()
+    }
+
+    // FIXME: support for EarlyUserAuthResult
+    pub fn read_request_from_server(&mut self, input: &[u8]) -> ConnectorResult<()> {
+        match self.state {
+            CredSSPState::CredsspInitial | CredSSPState::CredsspReplyNeeded => {
+                info!("read request from server: {:?}", input);
+                let message = credssp::TsRequest::from_buffer(input)
+                    .map_err(|e| reason_err!("CredSSP", "TsRequest decode: {e}"))?;
+                debug!(?message, "Received");
+                self.next_request = Some(message);
+                Ok(())
+            }
+            CredSSPState::CredsspEarlyUserAuthResult => {
+                let early_user_auth_result = credssp::EarlyUserAuthResult::from_buffer(input)
+                    .map_err(|e| custom_err!("credssp::EarlyUserAuthResult", e))?;
+
+                debug!(message = ?early_user_auth_result, "Received");
+
+                let credssp::EarlyUserAuthResult::Success = early_user_auth_result else {
+                    return Err(ConnectorError::new("CredSSP", ConnectorErrorKind::AccessDenied));
+                };
+                Ok(())
+            }
+            _ => Err(general_err!("CredSsp Sequence is Finished")),
+        }
+    }
+
+    pub fn process(&mut self) -> CredSspProcessGenerator<'_> {
+        let request = self.next_request.take().expect("next request");
+        info!("Ts request = {:?}", &request);
```

Comment:
With tracing, you can log values in a structured manner. Also this should not be `info` level. `debug` or `trace` is appropriate.
```suggestion
        let ts_request = self.next_request.take().expect("next request");
        debug!(?ts_request);
```